### PR TITLE
[evaluation] add automatic metric

### PR DESCRIPTION
**Pull request by Camille7777**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#3714 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
**Add/Modify files:
eval.py 
evaluator.py
metrics.py
requirements.txt**

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Chat/evaluate/metric.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+import jieba
+from nltk.translate.bleu_score import sentence_bleu
+from rouge_chinese import Rouge
+from bert_score import score
+from sklearn.metrics import precision_score, recall_score, f1_score
```
Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:5]: Python imports should follow PEP-8 standards.

Please refer this guide to reorganise your imports: https://composed.blog/python/imports

File: applications/Chat/evaluate/metric.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    for pred, target in zip(preds, targets):
+        pred_list = []
+        target_list = []
+        pred_list = (' '.join(jieba.cut(pred))).split()
+        target_list.append((' '.join(jieba.cut(target))).split())
+
+        bleu1 = sentence_bleu(target_list, pred_list, weights=(1, 0, 0, 0))
+        bleu2 = sentence_bleu(target_list, pred_list, weights=(0, 1, 0, 0))
+        bleu3 = sentence_bleu(target_list, pred_list, weights=(0, 0, 1, 0))
+        bleu4 = sentence_bleu(target_list, pred_list, weights=(0, 0, 0, 1))
```
Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:22]: As far as I understand, if you want to calculate 3-gram BLEU, you should consider 1-gram, 2-gram and 3-gram at the same time.

See the example here:
https://www.nltk.org/_modules/nltk/translate/bleu_score.html

Also you can also put weights in a list to avoid repetition code.

File: applications/Chat/evaluate/metric.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+import jieba
+from nltk.translate.bleu_score import sentence_bleu
+from rouge_chinese import Rouge
```
Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:3]: If this is only for Chinese one, your naming should show that this is Chinese Rouge. Please add alias for this import.

Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:3]: If those are external packages, you should add the packages in the installation process.

File: applications/Chat/evaluate/metric.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    bleu_score["bleu1"] = cumulate_bleu1 / len(preds)
+    bleu_score["bleu2"] = cumulate_bleu2 / len(preds)
+    bleu_score["bleu3"] = cumulate_bleu3 / len(preds)
+    bleu_score["bleu4"] = cumulate_bleu4 / len(preds)
+
+    return bleu_score
+
+
+def rouge_score(preds: list, targets: list) -> dict:
```
Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:37]: Rename to show that this is for Chinese?

File: applications/Chat/evaluate/metric.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
+import jieba
+from nltk.translate.bleu_score import sentence_bleu
+from rouge_chinese import Rouge
+from bert_score import score
```
Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:4]: same

File: applications/Chat/evaluate/metric.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        pred_list = [char for char in pred]
+        target_list = [char for char in target]
+
+        target_labels = [1] * len(target_list)
+        pred_labels = [int(pred_list[i] == target_list[i]) for i in range(0, len(target_list))]
+        precision_scores += precision_score(target_labels, pred_labels)
+
+    precision["precision"] = precision_scores / len(preds)
+
+    return precision
```
Comment (TongLi3701) [inline applications/Chat/evaluate/metric.py:116]: Are we calculating the precision by counting the number of overlaps between the `preds` and `targets`?

**Review by TongLi3701 — COMMENTED** [review]
Thanks Yuting.

Please see my comments above.

For each function, we should add docs to explain the purposes of the function, please see here as an example:

https://github.com/hpcaitech/ColossalAI/blob/main/applications/Chat/coati/models/llama/llama_actor.py

In addition, we should also add refs for non-common metrics like distinct.

Please address the comments and re-request review.

Thanks !

File: applications/Chat/evaluate/metric.py (hunk 6 above)
Comment (Camille7777) [reply applications/Chat/evaluate/metric.py:116]: for example, the target is "旅游", the pred is “旅行”, for now the whole calculation process will be:
target_list = ["旅", "游"]
pred_list = ["旅", "行"]
target_labels = [1, 1]
pred_labels = [1, 0]

then calculate for the precision score. This is mainly for classification and extraction categories.

File: applications/Chat/evaluate/metric.py (hunk 3 above)
Comment (Camille7777) [reply applications/Chat/evaluate/metric.py:3]: from rouge_chinese import Rouge as Rouge_cn

File: applications/Chat/evaluate/metric.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    """
+    bert_score = {"precision": 0, "recall": 0, "f1_score": 0}
+    pred_list = []
+    target_list = []
+
+    for pred, target in zip(preds, targets):
+        pred_list.append(' '.join(jieba.cut(pred)))
+        target_list.append(' '.join(jieba.cut(target)))
+    
+    P, R, F = score(pred_list, target_list, lang="zh", verbose=True)
```
Comment (chengeharrison) [inline applications/Chat/evaluate/metric.py:107]: Do we need to add another parameter `language` in this fuction and use an `if` to decide which parameter `lang` will take.

For example:
```python
if language == "cn":
    P, R, F = score(pred_list, target_list, lang="zh", verbose=True)
elif language == "en":
    P, R, F = score(pred_list, target_list, lang="en", verbose=True)
```

File: applications/Chat/requirements.txt [hunk 8]
```diff
@@ -11,3 +11,9 @@ sse_starlette
 wandb
 sentencepiece
 gpustat
+jieba
+bert-score
+rouge_chinese
+scikit-metrics
+nltk
+
```
Comment (TongLi3701) [inline applications/Chat/requirements.txt:19]: Remove additional space

**Review by TongLi3701 — APPROVED** [review]
Thanks Yuting.

File: applications/Chat/evaluate/metric.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+     ]
+
+    for pred, target in zip(preds, targets):
+        pred_list = (' '.join(jieba.cut(pred))).split()
+        target_list = [(' '.join(jieba.cut(target))).split()]
+
+        bleu = sentence_bleu(target_list, pred_list, weights = weights)
+        cumulative_bleu = [a+b for a, b in zip(cumulative_bleu, bleu)]
+
+    bleu_scores["bleu1"] = cumulative_bleu[0]/len(preds)
```
Comment (ver217) [inline applications/Chat/evaluate/metric.py:34]: Can you use a for-loop to simplify this?

Comment (ver217) [reply applications/Chat/evaluate/metric.py:34]: By the way, do you install pre-commit? It seems that the format is not correct.

Comment (Camille7777) [reply applications/Chat/evaluate/metric.py]:
```
for i in range(len(cumulative_bleu)):
        bleu_scores[f"bleu{i+1}"] = cumulative_bleu[i] / len(preds) 
```

File: applications/Chat/evaluate/eval.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        evaluator = Evaluator(metrics_per_category)
+        evaluator.evaluate(answers=answers, targets=targets)
+        evaluator.save(args.save_path)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--config_file', type=str, default=None, help='path to the file of target results')
+    parser.add_argument('--target_file', type=str, default=None, help='path to the file of target results')
+    parser.add_argument('--predict_file', type=str, default=None, help='path to the file of model prediction results')
```
Comment (chengeharrison) [inline applications/Chat/evaluate/eval.py:42]: The `predict_file` should be `target_folder` instead?

Comment (chengeharrison) [inline applications/Chat/evaluate/eval.py:42]: Should we use `predict_folder` instead? In this way we can evaluate all the models in one cml.

File: applications/Chat/evaluate/evaluator.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        self.params = params
+        self.stats = dict()
+
+    def battle(self, answers1: dict, answers2: dict) -> None:
+        """
+        Comparison between two models using GPT-4 as the reviewer.
+        """
+        pass
+
+    def evaluate(self, answers: dict, targets: dict) -> None:
```
Comment (chengeharrison) [inline applications/Chat/evaluate/evaluator.py:22]: The `answers` and `targets` should be `list` instead? `jload` fuction can return `dict` and `list` type.

If we use `predict_folder`, targets will be a dict like:
```python
{
    "name1": answers_list1,
    "name2": answers_list2,
    ......
}
```

File: applications/Chat/evaluate/eval.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            for metric_type, metrics in config["category"][category].items():
+                metrics_all[metric_type] = metrics
+            metrics_per_category[category] = metrics_all
+
+        answers = {}
+        targets = {}
+
+        with open(args.target_file, 'r', encoding='utf-8') as f1:
+            targets = json.load(f1)
+        f1.close()
```
Comment (ver217) [inline applications/Chat/evaluate/eval.py:26]: No need to close manually. The context will handle this.
